### PR TITLE
CP-11292: Remove storage motion constraints: Allow cross-pool copy for stopped VMs and templates

### PR DESCRIPTION
**Pull request by MihaelaStoica**

### CODE REVIEW

File: XenAdmin/Wizards/CrossPoolMigrateWizard/CrossPoolMigrateWizard.cs [hunk 1]
```diff
@@ -123,12 +127,33 @@ private bool IsIntraPoolMigration(KeyValuePair<string, VmMapping> mapping)
...
 
+        private bool IsIntraPoolMove()
+        {
+            return wizardMode == WizardMode.Move && m_vmMappings.All(IsIntraPoolMove);
+        }
+
+        private bool IsIntraPoolMove(KeyValuePair<string, VmMapping> mapping)
+        {
+            VM vm = xenConnection.Resolve(new XenRef<VM>(mapping.Key));
+            return vm.CanBeMoved && IsIntraPoolMigration(mapping);
```
Comment (GaborApatiNagy) [inline XenAdmin/Wizards/CrossPoolMigrateWizard/CrossPoolMigrateWizard.cs]: I think a null check would be useful here.

File: XenAdmin/Wizards/CrossPoolMigrateWizard/CrossPoolMigrateWizard.cs [hunk 2]
```diff
@@ -123,12 +127,33 @@ private bool IsIntraPoolMigration(KeyValuePair<string, VmMapping> mapping)
...
+
+        private bool IsCopyTemplate()
+        {
+            return wizardMode == WizardMode.Copy && m_vmMappings.Any(IsTemplate);
+        }
+
+        private bool IsTemplate(KeyValuePair<string, VmMapping> mapping)
+        {
+            VM vm = xenConnection.Resolve(new XenRef<VM>(mapping.Key));
+            return vm.is_a_template;
```
Comment (GaborApatiNagy) [inline XenAdmin/Wizards/CrossPoolMigrateWizard/CrossPoolMigrateWizard.cs]: I think a null check would be useful here.

File: XenModel/Actions/VM/VMCrossPoolMigrateAction.cs [hunk 3]
```diff
@@ -41,9 +40,10 @@ public class VMCrossPoolMigrateAction : PureAsyncAction
     {
         private readonly VmMapping mapping;
         private readonly XenAPI.Network transferNetwork;
+        private readonly bool copy;
 
-        public VMCrossPoolMigrateAction(VM vm, Host destinationHost, XenAPI.Network transferNetwork, VmMapping mapping)
-            : base(vm.Connection, GetTitle(vm, destinationHost))
+        public VMCrossPoolMigrateAction(VM vm, Host destinationHost, XenAPI.Network transferNetwork, VmMapping mapping, bool copy)
```
Comment (GaborApatiNagy) [inline XenModel/Actions/VM/VMCrossPoolMigrateAction.cs]: Bool copy does mean Copy or Migrate mode here. I think an Enum or a different action would be more clear here. (Especially copy = false does not suggest migrate to someone not familiar in this context)

Comment (stephen-turner) [inline XenModel/Actions/VM/VMCrossPoolMigrateAction.cs]: Could we just comment it to make it clearer, rather than make an extra function or enum?

**Comment by GaborApatiNagy** [comment]
Looks ok

**Comment by GaborApatiNagy** [comment]
:+1:

**Pull request merged**